go/apps/bulk_message/vumi_app.py

```diff
@@ -70,26 +84,62 @@ def send_message_via_window(self, conv, window_id, batch_id, to_addr,
     def process_command_bulk_send(self, user_account_key, conversation_key,
                                   batch_id, msg_options, content, dedupe,
                                   delivery_class, **extra_params):
-
+        """
+        Send a copy of a message to every contact in every group attached to
+        a conversation.
+
+        If this command is interrupted (by a worker restart, for example) the
+        next time it is processed it will avoid sending the message to contacts
+        that it has already been sent to. Note that the contact and opt-out
+        lookups still happen (because those are required for deduplication), so
+        there may be a lengthly delay before the remaining messages are sent if
+        the previous send was interrupted after a large number of messages.
+        """
         conv = yield self.get_conversation(user_account_key, conversation_key)
         if conv is None:
             log.warning("Cannot find conversation '%s' for user '%s'." % (
                 conversation_key, user_account_key))
             return
+        contact_store = conv.user_api.contact_store
+        addresses_seen = set()  # To deduplicate addresses, if asked.
 
-        to_addresses = []
-        for contacts_batch in (
-                yield conv.get_opted_in_contact_bunches(delivery_class)):
-            for contact in (yield contacts_batch):
-                to_addresses.append(contact.addr_for(delivery_class))
-        if dedupe:
-            to_addresses = set(to_addresses)
+        # NOTE: We fetch all contact keys before we start. This is suboptimal,
+        #       but we need the keys sorted and deduplicated.
+        contact_keys = yield conv.get_contact_keys()
```

Comment:
Yes. Changing that requires group contact listing code that returns sorted results, which is out of scope for this PR. (I can think of a way to do it, but it involves walking all relevant index/search result pages simultaneously (assuming search results are sorted by key) and that's a problem for search results if the result set is changing over the course of the query.)

We've seen in production that building a list of all contact keys before we start isn't as big a problem as loading all the contact objects to get the addresses, though.